File: app/src/main/java/scheffold/antoine/monitorsmartplugtp_link/util/TpLinkSmartPlugCommandsHelper.java
package scheffold.antoine.monitorsmartplugtp_link.util;

import androidx.annotation.NonNull;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

import scheffold.antoine.monitorsmartplugtp_link.Commands;

public class TpLinkSmartPlugCommandsHelper {

    public static final int TP_LINK_COMMUNICATION_PORT = 9999;

    public static int parseConsumingWatt(String toParse) {
        String[] seperatedByCommas = toParse.split(",");
        for (String str : seperatedByCommas) {
            if (str.contains("power_mw")) {
                String[] powerAndWatt = str.split(":");
                System.out.println(Arrays.toString(powerAndWatt));
                return Integer.parseInt(powerAndWatt[1]);
            }
        }
        throw new IllegalStateException("No Watt consumption in String");
    }

    public static String getSystemInfo(String ip) throws IOException {
        String cmd = Commands.COMMANDS.get("info");
        return queryCommand(ip, cmd);
    }

    public static String querySmartPlugMeter(String ip) throws IOException {
        String cmd = Commands.COMMANDS.get("energy");
        return queryCommand(ip, cmd);
    }

    public static String stopSmartPlug(String ip) throws IOException {
        String cmd = Commands.COMMANDS.get("off");
        return queryCommand(ip, cmd);
    }

    public static String startSmartPlug(String ip) throws IOException {
        String cmd = Commands.COMMANDS.get("on");
        return queryCommand(ip, cmd);
    }

    public static String configureWifi(String ip, String ssid, String password) throws IOException {
        String cmd = Commands.COMMANDS.get("configwifi");
        String formattedCmd = String.format(cmd, ssid, password);
        return queryCommand(ip, formattedCmd);
    }

    // TODO: 23.05.23 configure WIFI of smartplug

    @NonNull
    private static String queryCommand(String ip, String cmd) throws IOException {
        Socket socket = new Socket(ip, TP_LINK_COMMUNICATION_PORT);
        byte[] encryptedCmd = encrypt(cmd.getBytes());
        OutputStream socketOutputStream = socket.getOutputStream();
        socketOutputStream.write(encryptedCmd);
        socketOutputStream.flush();
        InputStream inputStream = socket.getInputStream();
        byte[] answer = new byte[2048];
        int a = inputStream.read(answer);
        socket.close();
        byte[] decryptedAnswer = decrypt(answer);
        byte[] smallerArray = cutOffArrayFromIndex(decryptedAnswer, a);
        return new String(smallerArray, StandardCharsets.UTF_8);
    }

    // Encryption and Decryption of TP-Link Smart Home Protocol
// XOR Autokey Cipher with starting key = 171
    private static byte[] decrypt (byte[] bytArr) {
        byte[] tmparr = new byte[bytArr.length];
        if (bytArr.length > 0) {
            int key = 171;
            for (int i = 4; i < bytArr.length; i++) {
                byte b = (byte) (key ^ bytArr[i]);
                key =  bytArr[i];
                tmparr[i] = b;
            }
        }
        return tmparr;
    }

    private static byte[] encrypt (byte[] bArr) {
        byte [] tmpArr = new byte[bArr.length + 4];
        byte[] otherArr = ByteBuffer.allocate(4).putInt(bArr.length).array();
        System.arraycopy(otherArr, 0, tmpArr, 0, otherArr.length);
        if (bArr.length > 0) {
            int key = 171;
            for (int i = 0; i < bArr.length; i++) {
                byte b = (byte) (key ^ bArr[i]);
                key =  b;
                tmpArr[i+4] = b;
            }
        }
        return tmpArr;
    }

    private static byte[] cutOffArrayFromIndex(byte[] decryptedAnswer, int length) {
        byte[] smallerArray = new byte[length];
        System.arraycopy(decryptedAnswer, 0, smallerArray, 0, length);
        return smallerArray;
    }


}

File: app/src/main/java/scheffold/antoine/monitorsmartplugtp_link/util/TpLinkSmartPlugCommandsHelper.kt
package scheffold.antoine.monitorsmartplugtp_link.util

import scheffold.antoine.monitorsmartplugtp_link.Commands
import java.io.IOException
import java.net.Socket
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.util.Arrays

object TpLinkSmartPlugCommandsHelper {
    const val TP_LINK_COMMUNICATION_PORT = 9999
    @JvmStatic
    fun parseConsumingWatt(toParse: String): Int {
        val seperatedByCommas =
            toParse.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (str in seperatedByCommas) {
            if (str.contains("power_mw")) {
                val powerAndWatt =
                    str.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                println(Arrays.toString(powerAndWatt))
                return powerAndWatt[1].toInt()
            }
        }
        throw IllegalStateException("No Watt consumption in String")
    }

    @Throws(IOException::class)
    fun getSystemInfo(ip: String): String {
        val cmd = Commands.COMMANDS["info"]
        return queryCommand(ip, cmd)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun querySmartPlugMeter(ip: String): String {
        val cmd = Commands.COMMANDS["energy"]
        return queryCommand(ip, cmd)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun stopSmartPlug(ip: String): String {
        val cmd = Commands.COMMANDS["off"]
        return queryCommand(ip, cmd)
    }

    @Throws(IOException::class)
    fun startSmartPlug(ip: String): String {
        val cmd = Commands.COMMANDS["on"]
        return queryCommand(ip, cmd)
    }

    @Throws(IOException::class)
    fun configureWifi(ip: String, ssid: String?, password: String?): String {
        val cmd = Commands.COMMANDS["configwifi"]
        val formattedCmd = String.format(cmd!!, ssid, password)
        return queryCommand(ip, formattedCmd)
    }

    // TODO: 23.05.23 configure WIFI of smartplug
    @Throws(IOException::class)
    private fun queryCommand(ip: String, cmd: String?): String {
        val socket = Socket(ip, TP_LINK_COMMUNICATION_PORT)
        val encryptedCmd = encrypt(cmd!!.toByteArray())
        val socketOutputStream = socket.getOutputStream()
        socketOutputStream.write(encryptedCmd)
        socketOutputStream.flush()
        val inputStream = socket.getInputStream()
        val answer = ByteArray(2048)
        val a = inputStream.read(answer)
        socket.close()
        val decryptedAnswer = decrypt(answer)
        val smallerArray = cutOffArrayFromIndex(decryptedAnswer, a)
        return String(smallerArray, StandardCharsets.UTF_8)
    }

    // Encryption and Decryption of TP-Link Smart Home Protocol
    // XOR Autokey Cipher with starting key = 171
    private fun decrypt(bytArr: ByteArray): ByteArray {
        val tmparr = ByteArray(bytArr.size)
        if (bytArr.size > 0) {
            var key = 171
            for (i in 4 until bytArr.size) {
                val b = (key xor bytArr[i].toInt()).toByte()
                key = bytArr[i].toInt()
                tmparr[i] = b
            }
        }
        return tmparr
    }

    private fun encrypt(bArr: ByteArray): ByteArray {
        val tmpArr = ByteArray(bArr.size + 4)
        val otherArr = ByteBuffer.allocate(4).putInt(bArr.size).array()
        System.arraycopy(otherArr, 0, tmpArr, 0, otherArr.size)
        if (bArr.size > 0) {
            var key = 171
            for (i in bArr.indices) {
                val b = (key xor bArr[i].toInt()).toByte()
                key = b.toInt()
                tmpArr[i + 4] = b
            }
        }
        return tmpArr
    }

    private fun cutOffArrayFromIndex(decryptedAnswer: ByteArray, length: Int): ByteArray {
        val smallerArray = ByteArray(length)
        System.arraycopy(decryptedAnswer, 0, smallerArray, 0, length)
        return smallerArray
    }
}
